File: Assets/Scripts/Levels/LevelValidator.cs
using System.Collections.Generic;
using UnityEngine;

namespace Sokoban.Levels
{
    public class ValidationIssue
    {
        public string Message;
        public ValidationIssue(string message) => Message = message;
    }

    /// <summary>Kiểm tra cấu trúc một màn. Không kiểm tra màn có giải được hay không.</summary>
    public static class LevelValidator
    {
        public static List<ValidationIssue> Validate(LevelData level)
        {
            var issues = new List<ValidationIssue>();
            if (level == null || level.rows == null || level.rows.Length == 0)
            {
                issues.Add(new ValidationIssue("Màn rỗng"));
                return issues;
            }

            int players = 0, boxes = 0, goals = 0;
            Vector2Int playerPos = default;

            for (int y = 0; y < level.rows.Length; y++)
            {
                string row = level.rows[y];
                for (int x = 0; x < row.Length; x++)
                {
                    char c = row[x];
                    if (c == SokobanChars.Player || c == SokobanChars.PlayerOnGoal)
                    {
                        players++;
                        playerPos = new Vector2Int(x, y);
                    }
                    if (c == SokobanChars.Box || c == SokobanChars.BoxOnGoal) boxes++;
                    if (c == SokobanChars.Goal || c == SokobanChars.BoxOnGoal ||
                        c == SokobanChars.PlayerOnGoal) goals++;
                }
            }

            if (players != 1)
                issues.Add(new ValidationIssue($"Phải có đúng một người chơi, đang có {players}"));
            if (boxes == 0)
                issues.Add(new ValidationIssue("Màn không có hộp nào"));
            else if (boxes != goals)
                issues.Add(new ValidationIssue($"Số hộp ({boxes}) khác số đích ({goals})"));

            if (players == 1 && !IsEnclosed(level, playerPos))
                issues.Add(new ValidationIssue("Vùng chơi chưa kín — người chơi đi ra ngoài lưới được"));

            return issues;
        }

        static bool IsEnclosed(LevelData level, Vector2Int start)
        {
            int height = level.rows.Length;
            var seen = new HashSet<Vector2Int> { start };
            var queue = new Queue<Vector2Int>();
            queue.Enqueue(start);

            var deltas = new[]
            {
                new Vector2Int(1, 0), new Vector2Int(-1, 0),
                new Vector2Int(0, 1), new Vector2Int(0, -1)
            };

            while (queue.Count > 0)
            {
                var p = queue.Dequeue();
                foreach (var d in deltas)
                {
                    var n = p + d;
                    if (n.y < 0 || n.y >= height) return false;

                    string row = level.rows[n.y];
                    if (n.x < 0 || n.x >= row.Length) return false;

                    if (row[n.x] == SokobanChars.Wall || seen.Contains(n)) continue;
                    seen.Add(n);
                    queue.Enqueue(n);
                }
            }

            return true;
        }
    }
}

File: web/src/levels/levelValidator.js
import { WALL, countPieces } from './sokobanChars.js';

/**
 * Checks the structure of a level. Does NOT check whether the level is solvable —
 * writing a Sokoban solver is a problem of its own, and out of scope.
 */
export function validateLevel(level) {
  if (!level || !level.rows || level.rows.length === 0) return ['Empty level'];

  const issues = [];
  const { players, boxes, goals, playerPos } = countPieces(level.rows);

  if (players !== 1) issues.push(`Must have exactly one player, found ${players}`);
  if (boxes === 0) issues.push('Level has no boxes');
  else if (boxes !== goals) issues.push(`Box count (${boxes}) differs from goal count (${goals})`);

  if (players === 1 && !isEnclosed(level, playerPos)) {
    issues.push('Play area is not sealed — the player can walk off the grid');
  }

  return issues;
}

/** Flood from the player; reaching outside the grid means the walls do not enclose it. */
function isEnclosed(level, start) {
  const height = level.rows.length;
  const seen = new Set([`${start.x},${start.y}`]);
  const queue = [start];
  const deltas = [[1, 0], [-1, 0], [0, 1], [0, -1]];

  while (queue.length > 0) {
    const p = queue.shift();
    for (const [dx, dy] of deltas) {
      const n = { x: p.x + dx, y: p.y + dy };
      if (n.y < 0 || n.y >= height) return false;

      const row = level.rows[n.y];
      if (n.x < 0 || n.x >= row.length) return false;

      const key = `${n.x},${n.y}`;
      if (row[n.x] === WALL || seen.has(key)) continue;
      seen.add(key);
      queue.push(n);
    }
  }

  return true;
}
